github/compliance/report.py

```diff
@@ -627,39 +627,85 @@ def _scanned_element_ctx_label(
 @functools.cache
 def _target_milestone(
     repo: github3.repos.Repository,
-    sprint: delivery.model.Sprint,
-):
+    sprints: tuple[delivery.model.Sprint],
+) -> tuple[
+    github3.repos.repo.milestone.Milestone | None,
+    list[github3.repos.repo.milestone.Milestone],
+]:
     return github.compliance.milestone.find_or_create_sprint_milestone(
         repo=repo,
-        sprint=sprint,
+        sprints=sprints,
     )
 
 
 @functools.cache
-def _target_sprint(
+def _upcoming_sprints(
+    delivery_svc_client: delivery.client.DeliveryServiceClient,
+    today: datetime.date=datetime.date.today(), # used to refresh the cache daily
+) -> list[delivery.model.Sprint]:
+    sprints = delivery_svc_client.sprints()
+    upcoming_sprints = [
+        sprint for sprint in sprints
+        if sprint.find_sprint_date(name='end_date').value.date() >= today
+    ]
+
+    if len(upcoming_sprints) == 0:
+        raise ValueError(f'no upcoming sprints found, all sprints ended before {today}')
+
+    return upcoming_sprints
+
+
+def _target_sprints(
     delivery_svc_client: delivery.client.DeliveryServiceClient,
     latest_processing_date: datetime.date=None,
     sprint_end_date: datetime.date=None,
-) -> delivery.model.Sprint:
+    sprints_count: int=1,
+) -> tuple[delivery.model.Sprint]:
     if not latest_processing_date and not sprint_end_date:
         raise ValueError(
           "At least one of 'latest_processing_date' and 'sprint_end_date' must not be 'None'"
         )
-    try:
-        if latest_processing_date:
-            target_sprint = delivery_svc_client.sprint_current(
-                before=latest_processing_date,
-                offset=-1,
-            )
-        elif sprint_end_date:
-            target_sprint = delivery_svc_client.sprint_current(
-                before=sprint_end_date,
-            )
-    except requests.HTTPError as http_error:
-        logger.warning(f'error determining tgt-sprint {http_error=} - falling back to current')
-        target_sprint = delivery_svc_client.sprint_current()
+    today = datetime.date.today()
+    sprints = _upcoming_sprints(
```

Comment:
As we rely on sprints being ordered by date, I would prefer to either explicitly order here (again), or at least add a comment / choose a variable name expressing that.

Reply:
The ordering of sprints is done a few lines below. The function `_upcoming_sprints` does not do any ordering.

Reply:
got it!